crates/uv-interpreter/src/managed/find.rs

```diff
@@ -1,47 +1,140 @@
+    /// but should not be relied on — instead the toolchains should be sorted later by
+    /// the parsed Python version.
+    fn find_all(&self) -> Result<impl DoubleEndedIterator<Item = Toolchain>, Error> {
+        let dirs = match fs_err::read_dir(&self.root) {
+            Ok(toolchain_dirs) => {
+                // Collect sorted directory paths; `read_dir` is not stable across platforms
+                let directories: BTreeSet<_> = toolchain_dirs
+                    .filter_map(|read_dir| match read_dir {
+                        Ok(entry) => match entry.file_type() {
+                            Ok(file_type) => file_type.is_dir().then_some(Ok(entry.path())),
+                            Err(err) => Some(Err(err)),
+                        },
+                        Err(err) => Some(Err(err)),
                     })
-                    .ok()
-            } else {
-                None
+                    .collect::<Result<_, std::io::Error>>()
+                    .map_err(|err| Error::ReadError {
+                        dir: self.root.clone(),
+                        err,
+                    })?;
+                directories
             }
-        });
+            Err(err) if err.kind() == std::io::ErrorKind::NotFound => BTreeSet::default(),
+            Err(err) => {
+                return Err(Error::ReadError {
+                    dir: self.root.clone(),
+                    err,
+                })
+            }
+        };
+        Ok(dirs
+            .into_iter()
+            .map(|path| Toolchain::new(path).unwrap())
+            .rev())
+    }
+
+    /// Iterate over toolchains that support the current platform.
+    pub fn find_matching_current_platform(
+        &self,
+    ) -> Result<impl DoubleEndedIterator<Item = Toolchain>, Error> {
+        let platform_key = platform_key_from_env()?;
+
+        let iter = InstalledToolchains::from_settings()?
+            .find_all()?
+            .filter(move |toolchain| {
+                toolchain
+                    .path
+                    .file_name()
+                    .map(OsStr::to_string_lossy)
+                    .is_some_and(|filename| filename.ends_with(&platform_key))
+            });
+
+        Ok(iter)
+    }
 
-    Ok(iter)
+    /// Iterate over toolchains that satisfy the given Python version on this platform.
+    ///
+    /// ## Errors
+    ///
+    /// - The platform metadata cannot be read
+    /// - A directory in the toolchain directory cannot be read
+    pub fn find_version<'a>(
+        &self,
+        version: &'a PythonVersion,
+    ) -> Result<impl DoubleEndedIterator<Item = Toolchain> + 'a, Error> {
+        Ok(self
+            .find_matching_current_platform()?
+            .filter(move |toolchain| {
+                toolchain
+                    .path
+                    .file_name()
+                    .map(OsStr::to_string_lossy)
+                    .is_some_and(|filename| filename.starts_with(&format!("cpython-{version}")))
```

Comment:
You could do something like:

```diff
diff --git a/crates/uv-interpreter/src/managed/find.rs b/crates/uv-interpreter/src/managed/find.rs
index fd15c484..40f609cd 100644
--- a/crates/uv-interpreter/src/managed/find.rs
+++ b/crates/uv-interpreter/src/managed/find.rs
@@ -128,7 +128,11 @@ impl InstalledToolchains {
                     .path
                     .file_name()
                     .map(OsStr::to_string_lossy)
-                    .is_some_and(|filename| filename.starts_with(&format!("cpython-{version}")))
+                    .is_some_and(|filename| {
+                        filename
+                            .strip_prefix("cpython-")
+                            .is_some_and(|prefix| prefix.starts_with(&version.to_string()))
+                    })
             }))
     }
```

That would avoid allocation for anything that doesn't start with `cpython-`. But, IDK, perhaps this already gets optimized out anyway.

Reply:
What you have now is more readable, I'd leave it as-is.

Reply:
Probably going to parse the folder name into metadata _once_ when constructing the `Toolchain` in the future anyway